File: src/bmsx_cpp/render/3d/mesh/rom_source.cpp
#include "render/3d/mesh/rom_source.h"

#include "rompack/tokens.h"

namespace bmsx {
namespace {

AssetToken meshAssetToken(const GameView::VdpMeshRenderEntry& entry) {
	return makeAssetToken(entry.modelTokenLo, entry.modelTokenHi);
}

} // namespace

bool hasMeshRomDrawSources(const RuntimeRomPackage& rom, const GameView& view) {
	if (view.vdpMeshCount == 0u) {
		return false;
	}
	for (size_t index = 0u; index < view.vdpMeshCount; ++index) {
		const GameView::VdpMeshRenderEntry& entry = view.vdpMeshes[index];
		const auto it = rom.model.find(meshAssetToken(entry));
		if (it == rom.model.end() || entry.meshIndex >= it->second.meshes.size()) {
			return false;
		}
	}
	return true;
}

MeshRomDrawSource resolveMeshRomDrawSource(const RuntimeRomPackage& rom, const GameView::VdpMeshRenderEntry& entry) {
	const ModelAsset& model = rom.model.find(meshAssetToken(entry))->second;
	return {model, model.meshes[entry.meshIndex]};
}

} // namespace bmsx

File: src/bmsx/render/3d/mesh/rom_source.ts
import type { Runtime } from '../../../machine/runtime/runtime';
import type { GLTFMesh, GLTFModel } from '../../../rompack/format';
import type { GameView } from '../../gameview';

export interface MeshRomDrawSource {
	model: GLTFModel;
	mesh: GLTFMesh;
}

const meshRomDrawSource: MeshRomDrawSource = {
	model: null,
	mesh: null,
};

export function hasMeshRomDrawSources(runtime: Runtime, view: GameView): boolean {
	const meshCount = view.vdpMeshCount;
	if (meshCount === 0) {
		return false;
	}
	for (let entryIndex = 0; entryIndex < meshCount; entryIndex += 1) {
		const tokenHiModels = runtime.activePackage.modelByToken.get(view.vdpMeshModelTokenHi[entryIndex]);
		if (!tokenHiModels) {
			return false;
		}
		const model = tokenHiModels.get(view.vdpMeshModelTokenLo[entryIndex]);
		if (!model || !model.meshes[view.vdpMeshIndex[entryIndex]]) {
			return false;
		}
	}
	return true;
}

export function resolveMeshRomDrawSource(runtime: Runtime, view: GameView, entryIndex: number): MeshRomDrawSource {
	const tokenHiModels = runtime.activePackage.modelByToken.get(view.vdpMeshModelTokenHi[entryIndex]);
	const model = tokenHiModels.get(view.vdpMeshModelTokenLo[entryIndex]);
	const meshIndex = view.vdpMeshIndex[entryIndex];
	const mesh = model.meshes[meshIndex];
	meshRomDrawSource.model = model;
	meshRomDrawSource.mesh = mesh;
	return meshRomDrawSource;
}
